File: pset6/credit/credit.py
from cs50 import get_string


def main():
    n = get_string("Number:")
    check(n)


def check(n):
    length = len(n)

    # check if there are 13~16 digit numbers
    if length < 13 or length > 16:
        print("INVALID")
        return 1

    # check American Express
    if (length == 15 and (n[:2] == '34' or n[:2] == '37')):
        if (algorithm(n) == 0):
            print("AMEX")
            return 0
        else:
            print("INVALID")
            return 1

    # check MasterCard
    elif (length == 16 and (int(n[:2]) >= 51 and int(n[:2]) <= 55)):
        if (algorithm(n) == 0):
            print("MASTERCARD")
            return 0
        else:
            print("INVALID")
            return 1

    # check VISA
    elif n[0] == '4':
        if (algorithm(n) == 0):
            print("VISA")
            return 0
        else:
            print("INVALID")
            return 1

    else:
        print("INVALID")
        return 1


def algorithm(n):
    numbers = list(n)
    numbers = [int(x) for x in numbers]
    flag = 0
    sum = 0
    for j in range(len(n)-1, -1, -1):
        if flag == 0:
            sum += numbers[j]
            flag = 1
        elif flag == 1:
            temp = numbers[j] * 2
            while temp > 0:
                sum += temp % 10
                temp //= 10
            flag = 0

    if sum % 10 == 0:
        return 0
    else:
        return 1


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int check(long n);
int algorithm(int n[], int i);

int main(void)
{
    long n = get_long("Number:");
    while (n <= 0)
    {
        n = get_long("Number:");
    }
    check(n);
}

int check(long n)
{
    int number[16] = {0};
    int length = 0;
    
    // check if there are 13~16 digit numbers
    if (n < 1000000000000 || n >= 10000000000000000)
    {
        printf("INVALID\n");
        return 1;
    }
    
    // get each digit number
    while (n != 0)
    {
        number[length] = n % 10;
        n /= 10;
        length++;
    }
    length--;

    
    // count the first two numbers
    int head = number[length] * 10 + number[length - 1];
    
    
    // check American Express
    if (length == 14 && (head == 34 || head == 37))
    {
        if (algorithm(number, length) == 0)
        {
            printf("AMEX\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 1;
        }
    }
    
    // check MasterCard
    else if (length == 15 && (head >= 51 && head <= 55))
    {
        if (algorithm(number, length) == 0)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 1;
        }
    }
    
    // check VISA
    else if ((length >= 12 && length <= 15) && number[length] == 4)
    {
        if (algorithm(number, length) == 0)
        {
            printf("VISA\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 1;
        }
        
    }
    
    else
    {
        printf("INVALID\n");
        return 1;
    }
}

int algorithm(int n[], int i)
{
    int flag = 0;
    int sum = 0;
    for (int j = 0; j <= i; j++)
    {
        if (flag == 0)
        {
            sum += n[j];
            flag = 1;
        }
        else if (flag == 1)
        {
            int temp = n[j] * 2;
            while (temp > 0)
            {
                sum += temp % 10;
                temp /= 10;
            }
            flag = 0;
        }
    }
    
    if (sum % 10 == 0)
    {
        return 0;
    }
    else
    {
        return 1;
    }
}

